File: tests/structurefunction_nc_test.cc
//
// APFEL++ 2017
//
// Author: Valerio Bertone: valerio.bertone@cern.ch
//

#include <apfel/apfelxx.h>

#include <iomanip>

int main()
{
  //SetVerbosityLevel(0);
  apfel::Banner();

  // x-space grid
  const apfel::Grid g{{apfel::SubGrid{100,1e-5,3}, apfel::SubGrid{60,1e-1,3}, apfel::SubGrid{50,6e-1,3}, apfel::SubGrid{50,8e-1,3}}};

  // Initial scale
  const double mu0 = sqrt(2);

  // Vectors of thresholds
  const std::vector<double> Thresholds = {0, 0, 0, sqrt(2), 4.5, 175};

  // Perturbative order
  const int PerturbativeOrder = 2;

  // Running coupling
  apfel::AlphaQCD a{0.35, sqrt(2), Thresholds, PerturbativeOrder};
  const apfel::TabulateObject<double> Alphas{a, 100, 0.9, 1001, 3};
  const auto as = [&] (double const& mu) -> double{ return Alphas.Evaluate(mu); };

  // Effective charges
  std::function<std::vector<double>(double const&)> fBq = [=] (double const& Q) -> std::vector<double> { return apfel::ElectroWeakCharges(Q, false); };
  std::function<std::vector<double>(double const&)> fDq = [=] (double const& Q) -> std::vector<double> { return apfel::ParityViolatingElectroWeakCharges(Q, false); };

  // Initialize QCD evolution objects
  const auto DglapObj = InitializeDglapObjectsQCD(g, Thresholds);

  // Construct the DGLAP object
  auto EvolvedPDFs = BuildDglap(DglapObj, apfel::LHToyPDFs, mu0, PerturbativeOrder, as);

  // Tabulate PDFs
  const apfel::TabulateObject<apfel::Set<apfel::Distribution>> TabulatedPDFs{*EvolvedPDFs, 50, 1, 1000, 3};

  // Evolved PDFs
  const auto PDFs = [&] (double const& x, double const& Q) -> std::map<int, double> { return TabulatedPDFs.EvaluateMapxQ(x, Q); };

  // Initialize coefficient functions
  const auto F2Obj = InitializeF2NCObjectsZM(g, Thresholds);
  const auto FLObj = InitializeFLNCObjectsZM(g, Thresholds);
  const auto F3Obj = InitializeF3NCObjectsZM(g, Thresholds);

  // Initialize structure functions
  const auto F2 = BuildStructureFunctions(F2Obj, PDFs, PerturbativeOrder, as, fBq);
  const auto FL = BuildStructureFunctions(FLObj, PDFs, PerturbativeOrder, as, fBq);
  const auto F3 = BuildStructureFunctions(F3Obj, PDFs, PerturbativeOrder, as, fDq);

  // Tabulate Structure functions
  const apfel::TabulateObject<apfel::Distribution> F2total {[&] (double const& Q) -> apfel::Distribution{ return F2.at(0).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F2light {[&] (double const& Q) -> apfel::Distribution{ return F2.at(1).Evaluate(Q) + F2.at(2).Evaluate(Q) + F2.at(3).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F2charm {[&] (double const& Q) -> apfel::Distribution{ return F2.at(4).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F2bottom{[&] (double const& Q) -> apfel::Distribution{ return F2.at(5).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};

  const apfel::TabulateObject<apfel::Distribution> FLtotal {[&] (double const& Q) -> apfel::Distribution{ return FL.at(0).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> FLlight {[&] (double const& Q) -> apfel::Distribution{ return FL.at(1).Evaluate(Q) + FL.at(2).Evaluate(Q) + FL.at(3).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> FLcharm {[&] (double const& Q) -> apfel::Distribution{ return FL.at(4).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> FLbottom{[&] (double const& Q) -> apfel::Distribution{ return FL.at(5).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};

  const apfel::TabulateObject<apfel::Distribution> F3total {[&] (double const& Q) -> apfel::Distribution{ return F3.at(0).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F3light {[&] (double const& Q) -> apfel::Distribution{ return F3.at(1).Evaluate(Q) + F3.at(2).Evaluate(Q) + F3.at(3).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F3charm {[&] (double const& Q) -> apfel::Distribution{ return F3.at(4).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};
  const apfel::TabulateObject<apfel::Distribution> F3bottom{[&] (double const& Q) -> apfel::Distribution{ return F3.at(5).Evaluate(Q); }, 50, 1, 200, 3, Thresholds};

  apfel::Timer t;

  // Final scale
  const double Q = 100;

  std::cout << std::scientific << std::endl;
  std::cout << "Alphas(Q) = " << as(Q) << std::endl;
  std::cout << std::endl;

  // Print results
  const std::vector<double> xlha = {1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 3e-1, 5e-1, 7e-1, 9e-1};

  std::cout << "    x   "
            << "  F2light   "
            << "  F2charm   "
            << "  F2bottom  "
            << "  F2total   "
            << std::endl;
  for (double const& x : xlha)
    std::cout << std::setprecision(1) << x << "  " << std::setprecision(4)
              << F2light.EvaluatexQ(x, Q)  << "  "
              << F2charm.EvaluatexQ(x, Q)  << "  "
              << F2bottom.EvaluatexQ(x, Q) << "  "
              << F2total.EvaluatexQ(x, Q)  << "  "
              << std::endl;
  std::cout << std::endl;

  std::cout << "    x   "
            << "  FLlight   "
            << "  FLcharm   "
            << "  FLbottom  "
            << "  FLtotal   "
            << std::endl;
  for (double const& x : xlha)
    std::cout << std::setprecision(1) << x << "  " << std::setprecision(4)
              << FLlight.EvaluatexQ(x, Q)  << "  "
              << FLcharm.EvaluatexQ(x, Q)  << "  "
              << FLbottom.EvaluatexQ(x, Q) << "  "
              << FLtotal.EvaluatexQ(x, Q)  << "  "
              << std::endl;
  std::cout << std::endl;

  std::cout << "    x   "
            << "  F3light   "
            << "  F3charm   "
            << "  F3bottom  "
            << "  F3total   "
            << std::endl;
  for (double const& x : xlha)
    std::cout << std::setprecision(1) << x << "  " << std::setprecision(4)
              << F3light.EvaluatexQ(x, Q)  << "  "
              << F3charm.EvaluatexQ(x, Q)  << "  "
              << F3bottom.EvaluatexQ(x, Q) << "  "
              << F3total.EvaluatexQ(x, Q)  << "  "
              << std::endl;
  std::cout << std::endl;

  t.stop();

  const int k = 1000000;
  std::cout << "Interpolating " << k << " times F2 on the grid... ";
  t.start();
  for (int i = 0; i < k; i++)
    F2total.EvaluatexQ(0.05, Q);
  t.stop();

  return 0;
}

File: pywrap/structurefunction_nc_test.py
import apfelpy as ap
import numpy as np

# Show banner
ap.Banner()

# x-space grid
g = ap.Grid([ap.SubGrid(100, 1e-5, 3), ap.SubGrid(60, 1e-1, 3), ap.SubGrid(50, 6e-1, 3), ap.SubGrid(50, 8e-1, 3)])

# Initial scale
mu0 = np.sqrt(2)

# Vectors of masses and thresholds
Thresholds = [0, 0, 0, np.sqrt(2), 4.5, 175]

# Perturbative order
PerturbativeOrder = 2

# Running coupling
a = ap.AlphaQCD(0.35, np.sqrt(2), Thresholds, PerturbativeOrder)
Alphas = ap.TabulateObject(a, 100, 0.9, 1001, 3)

# Effective charges
def fBq(Q):
    return ap.utilities.ElectroWeakCharges(Q, False)

def fDq(Q):
    return ap.utilities.ParityViolatingElectroWeakCharges(Q, False)

# Initialize QCD evolution objects
DglapObj = ap.initializers.InitializeDglapObjectsQCD(g, Thresholds)

# Construct the DGLAP objects
EvolvedPDFs = ap.builders.BuildDglap(DglapObj, ap.utilities.LHToyPDFs, mu0, PerturbativeOrder, Alphas.Evaluate)

# Tabulate PDFs
TabulatedPDFs = ap.TabulateObjectSetD(EvolvedPDFs, 50, 1, 1000, 3)

# Initialize coefficient functions
F2Obj = ap.initializers.InitializeF2NCObjectsZM(g, Thresholds)
FLObj = ap.initializers.InitializeFLNCObjectsZM(g, Thresholds)
F3Obj = ap.initializers.InitializeF3NCObjectsZM(g, Thresholds)

# Initialize structure functions
F2 = ap.builders.BuildStructureFunctions(F2Obj, TabulatedPDFs.EvaluateMapxQ, PerturbativeOrder, Alphas.Evaluate, fBq)
FL = ap.builders.BuildStructureFunctions(FLObj, TabulatedPDFs.EvaluateMapxQ, PerturbativeOrder, Alphas.Evaluate, fBq)
F3 = ap.builders.BuildStructureFunctions(F3Obj, TabulatedPDFs.EvaluateMapxQ, PerturbativeOrder, Alphas.Evaluate, fDq)

# Tabulate Structure functions
F2total  = ap.TabulateObjectD(F2[0].Evaluate, 50, 1, 200, 3, Thresholds)
F2light  = ap.TabulateObjectD(lambda Q: F2[1].Evaluate(Q) + F2[2].Evaluate(Q) + F2[3].Evaluate(Q), 50, 1, 200, 3, Thresholds)
F2charm  = ap.TabulateObjectD(F2[4].Evaluate, 50, 1, 200, 3, Thresholds)
F2bottom = ap.TabulateObjectD(F2[5].Evaluate, 50, 1, 200, 3, Thresholds)

FLtotal  = ap.TabulateObjectD(FL[0].Evaluate, 50, 1, 200, 3, Thresholds)
FLlight  = ap.TabulateObjectD(lambda Q: FL[1].Evaluate(Q) + FL[2].Evaluate(Q) + FL[3].Evaluate(Q), 50, 1, 200, 3, Thresholds)
FLcharm  = ap.TabulateObjectD(FL[4].Evaluate, 50, 1, 200, 3, Thresholds)
FLbottom = ap.TabulateObjectD(FL[5].Evaluate, 50, 1, 200, 3, Thresholds)

F3total  = ap.TabulateObjectD(F3[0].Evaluate, 50, 1, 200, 3, Thresholds)
F3light  = ap.TabulateObjectD(lambda Q: F3[1].Evaluate(Q) + F3[2].Evaluate(Q) + F3[3].Evaluate(Q), 50, 1, 200, 3, Thresholds)
F3charm  = ap.TabulateObjectD(F3[4].Evaluate, 50, 1, 200, 3, Thresholds)
F3bottom = ap.TabulateObjectD(F3[5].Evaluate, 50, 1, 200, 3, Thresholds)

# Timer
t = ap.Timer()

# Final scale
Q = 100

# Print results
xlha = [1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 3e-1, 5e-1, 7e-1, 9e-1]
print("\nAlphaQCD(Q) = " , Alphas.Evaluate(Q))

print("\n   x    "
          , "  F2light   "
          , "  F2charm   "
          , "  F2bottom  "
          , "  F2total   "
         )
for x in xlha:
    print(format(x, ".1e"), " ",
              format(F2light.EvaluatexQ(x, Q), ".4e"), " ",
              format(F2charm.EvaluatexQ(x, Q), ".4e"), " ",
              format(F2bottom.EvaluatexQ(x, Q), ".4e"), " ",
              format(F2total.EvaluatexQ(x, Q), ".4e"))

print("\n   x    "
          , "  FLlight   "
          , "  FLcharm   "
          , "  FLbottom  "
          , "  FLtotal   "
         )
for x in xlha:
    print(format(x, ".1e"), " ",
              format(FLlight.EvaluatexQ(x, Q), ".4e"), " ",
              format(FLcharm.EvaluatexQ(x, Q), ".4e"), " ",
              format(FLbottom.EvaluatexQ(x, Q), ".4e"), " ",
              format(FLtotal.EvaluatexQ(x, Q), ".4e"))

print("\n   x    "
          , "  F3light   "
          , "  F3charm   "
          , "  F3bottom  "
          , "  F3total   "
         )
for x in xlha:
    print(format(x, ".1e"), " ",
              format(F3light.EvaluatexQ(x, Q), ".4e"), " ",
              format(F3charm.EvaluatexQ(x, Q), ".4e"), " ",
              format(F3bottom.EvaluatexQ(x, Q), ".4e"), " ",
              format(F3total.EvaluatexQ(x, Q), ".4e"))

t.start()

k = 1000000
print("\nInterpolating ", k, " times a single F2 on the grid... ")
t.start()
for i in range(k):
  F2total.EvaluatexQ(0.05, Q)
t.stop()
